File: backend/app/schemas/alert.py
from datetime import datetime
from uuid import UUID

from pydantic import BaseModel

from app.schemas.event import EventResponse


class AlertResponse(BaseModel):
    id: UUID
    host_id: UUID
    rule_id: UUID | None
    severity: str
    title: str
    description: str | None
    status: str
    confidence: float | None = None
    assigned_to: UUID | None = None
    mitre_technique_id: str | None = None
    created_at: datetime
    resolved_at: datetime | None
    resolved_by: UUID | None

    model_config = {"from_attributes": True}


class AlertListResponse(BaseModel):
    items: list[AlertResponse]
    total: int
    page: int
    page_size: int


class AlertStatusUpdate(BaseModel):
    status: str
    assigned_to: UUID | None = None


class AlertBulkUpdate(BaseModel):
    alert_ids: list[UUID]
    status: str | None = None
    assigned_to: UUID | None = None


class AlertBulkUpdateResponse(BaseModel):
    updated: int
    not_found: list[UUID]


class AlertInvestigationHost(BaseModel):
    id: UUID
    name: str
    hostname: str | None
    status: str
    ip_address: str | None
    risk_score: int | None = None


class AlertInvestigationTimeline(BaseModel):
    id: UUID
    title: str
    severity: str
    confidence: float
    started_at: datetime
    status: str


class AlertInvestigationResponse(BaseModel):
    alert: AlertResponse
    host: AlertInvestigationHost
    events: list[EventResponse]
    timelines: list[AlertInvestigationTimeline]

File: frontend/lib/types/alert.ts
/**
 * Canonical alert types matching the backend AlertResponse schema.
 * All alert-related components should import from here.
 */

export interface Alert {
  id: string;
  host_id: string;
  rule_id: string | null;
  source: string | null;
  mitre_technique_id: string | null;
  mitre_tactic: string | null;
  confidence: number | null;
  severity: string;
  title: string;
  description: string | null;
  status: string;
  created_at: string;
  resolved_at: string | null;
  resolved_by: string | null;
  assigned_to: string | null;
  feedback_label: string | null;
  feedback_note: string | null;
  feedback_at: string | null;
  feedback_by: string | null;
}

export interface AlertListResponse {
  items: Alert[];
  total: number;
  page: number;
  page_size: number;
}

export interface AlertHost {
  id: string;
  name: string;
  hostname: string | null;
  status: string;
  ip_address: string | null;
  risk_score: number | null;
}

export interface AlertEvent {
  id: string;
  event_type: string;
  severity: string;
  description: string | null;
  timestamp: string;
}

export interface AlertTimeline {
  id: string;
  title: string;
  severity: string;
  confidence: number;
  started_at: string;
  status: string;
}

export interface InvestigationData {
  alert: Alert;
  host: AlertHost;
  events: AlertEvent[];
  timelines: AlertTimeline[];
}

export interface AlertAiSummary {
  alert_id: string;
  summary: string;
  investigation_steps: string[];
  recommended_actions: string[];
  provider: string;
}

export interface AlertFilters {
  status: string;
  severity: string;
  host_id: string;
  rule_name: string;
  q: string;
  mitre_technique_id: string;
}

export const ALERT_STATUSES = ["open", "investigating", "resolved", "closed"] as const;
export const ALERT_SEVERITIES = ["critical", "high", "medium", "low"] as const;

export type AlertStatus = (typeof ALERT_STATUSES)[number];
